Mathlib/Tactic/Linter/EmptyLine.lean

```diff
@@ -0,0 +1,153 @@
+It produces the array of the syntax nodes in `stx` where `f` is `true`.
+
+See also the similar `filterMap` and `filterMapM`.
+-/
+def filter (stx : Syntax) (f : Syntax → Bool) : Array Syntax :=
+  stx.filterMap fun s => if f s then some s else none
+
+end Lean.Syntax
+
+namespace Mathlib.Linter
+
+/--
+The "emptyLine" linter emits a warning on empty lines inside a command, but outside of a
+doc-string/module-doc.
+
+The linter is only active when there are no other warning, so as to not add noise when developing
+incomplete proofs.
+-/
+register_option linter.style.emptyLine : Bool := {
+  defValue := false
+  descr := "enable the emptyLine linter"
+}
+
+namespace EmptyLine
+
+/-- The `SyntaxNodeKind`s where the `EmptyLine` linter is not active -/
+abbrev AllowEmptyLines : Std.HashSet SyntaxNodeKind := Std.HashSet.emptyWithCapacity
+  |>.insert ``Parser.Command.docComment
+  |>.insert ``Parser.Command.moduleDoc
+  |>.insert ``Parser.Command.mutual
+  |>.insert `str
+
+/-- If a file contains one of these names as segments, we disable the `emptyLine` linter. -/
+abbrev SkippedFileSegments : Std.HashSet Name := Std.HashSet.emptyWithCapacity
+  |>.insert `Tactic
+  |>.insert `Util
+  |>.insert `Meta
+
+@[inherit_doc Mathlib.Linter.linter.style.emptyLine]
+def emptyLineLinter : Linter where run := withSetOptionIn fun stx ↦ do
+  unless Linter.getLinterValue linter.style.emptyLine (← getLinterOptions) do
+    return
+  -- The linter does not report anything on incomplete proofs, e.g. proofs containing `sorry`
+  -- or `stop`.
+  if (← get).messages.reportedPlusUnreported.any (!·.severity matches .information) then
+    return
+  if !((← getMainModule).components.filter SkippedFileSegments.contains).isEmpty then
+    return
+  -- We ignore empty lines "after" the command finished.
+  let stx := stx.unsetTrailing
+  let allowed := stx.filter (AllowEmptyLines.contains ·.getKind)
+  let allowedRanges := allowed.filterMap (·.getRange?)
+  if let some str := stx.getSubstring? then
+    if let one::rest := str.toString.trimRight.splitOn "\n\n" then
+      if rest.isEmpty then return
+      -- We extract all trailing ranges of all syntax nodes in `stx`, after we remove
+      -- leading and trailing whitespace from them.
+      -- These ranges typically represent embedded comments and we ignore line breaks inside them.
+      -- We do inspect leading and trailing whitespace though.
+      -- We treat `where` specially, since we allow empty lines in `where` fields.
+      let trails := stx.filterMap fun s =>
+        if let some str := s.getTrailing?
+        then
+          -- Handle `where` and `where` fields.
+          if s.getAtomVal == "where" ||
+             s.isOfKind ``Parser.Term.structInstField ||
+             s.isOfKind ``Parser.Command.structSimpleBinder then
+            s.getTrailing?.map (·.getRange)
+          else
+            let strim := str.trim
+            if (strim.splitOn "\n\n").length != 1 then
```

Comment:
There's no need to do a full split if you only care whether there are any matches; a find would be faster.